app/kubemci/pkg/gcp/sslcert/sslcertsyncer_test.go

```diff
@@ -100,25 +105,70 @@ func TestEnsureSSLCert(t *testing.T) {
 	}
 }
 
-func setupIng() (*v1beta1.Ingress, kubeclient.Interface) {
+func TestEnsureSSLCertForPreShared(t *testing.T) {
+	lbName := "lb-name"
+	certName := "testCert"
+	scp := ingresslb.NewFakeLoadBalancers("" /* name */, nil /* namer */)
+	namer := utilsnamer.NewNamer("mci1", lbName)
+	scs := NewSSLCertSyncer(namer, scp)
+	// GET should return NotFound.
+	if _, err := scp.GetSslCertificate(certName); err == nil {
+		t.Fatalf("expected NotFound error, got nil")
+	}
+	ing, client := setupIng(false /* withSecret */)
+
+	// EnsureSSLCert should give an error when both the secret and pre shared annotation are missing.
+	_, err := scs.EnsureSSLCert(lbName, ing, client, false /*forceUpdate*/)
+	if err == nil {
+		t.Errorf("unexpected nil error, expected error since both secret and pre shared cert annotation are missing")
+	}
+	// Adding the annotation without a cert should still return an error.
+	ing.ObjectMeta.Annotations = map[string]string{
+		annotations.PreSharedCertKey: certName,
+	}
+	_, err = scs.EnsureSSLCert(lbName, ing, client, false /*forceUpdate*/)
+	if err == nil {
+		t.Errorf("unexpected nil error, expected error since the pre shared cert is missing")
+	}
+
+	// EnsureSSLCert should return success if the cert exists as expected.
+	if _, err := scp.CreateSslCertificate(&compute.SslCertificate{
+		Name: certName,
+	}); err != nil {
+		t.Errorf("unexpected error in creating ssl cert: %s", err)
+	}
+	ensuredCertName, err := scs.EnsureSSLCert(lbName, ing, client, false /*forceUpdate*/)
+	if err != nil {
+		t.Errorf("unexpected error in ensuring ssl cert: %s", err)
+	}
+	if ensuredCertName != certName {
+		t.Errorf("unexpected ensured cert name, expected: %s, actual: %s", certName, ensuredCertName)
+	}
+}
+
+func setupIng(withSecret bool) (*v1beta1.Ingress, kubeclient.Interface) {
 	// Create ingress with secret for SSL cert.
 	svcName := "svcName"
 	svcPort := "svcPort"
 	secretName := "certSecret"
 	ing := &v1beta1.Ingress{
 		Spec: v1beta1.IngressSpec{
-			TLS: []v1beta1.IngressTLS{
-				{
-					SecretName: secretName,
-				},
-			},
 			Backend: &v1beta1.IngressBackend{
 				ServiceName: svcName,
 				ServicePort: intstr.FromString(svcPort),
 			},
 		},
 	}
 	client := &fake.Clientset{}
+	if !withSecret {
```

Comment:
nit: I usually like early returns, but this feels like it's setting us up for someone to forget that there was an early return, and add code at the end of the function that doesn't always run when it should. thoughts?

Reply:
right. Refactored the secret part into a separate method